### PR TITLE
Add systemd setup that supports workers

### PR DESCRIPTION
**Pull request by targodan**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://github.com/matrix-org/synapse/blob/master/CONTRIBUTING.rst#changelog)
* [x] Pull request includes a [sign off](https://github.com/matrix-
org/synapse/blob/master/CONTRIBUTING.rst#sign-off)

### Content

The currently distributed setup in the debian package does not work in
conjunction with workers. This setup provides a way to manage workers
with systemd. It does however not require workers. You can use this
setup without workers,  you just haveto make sure that the homeserver
is forking and writes its PID file to the location the service is looking in.

It also supplies a nice way to autostart services like bridges alongside synapse
utilizing its own `matrix.target`.

Should the package manager for the debian package be interested in this
and/or has any questions about this, feel free to comment here or contact me
via matrix @luca:targodan.de

Signed-off-by: Luca Corbatto <luca@corbatto.de>

### CODE REVIEW

**Comment by targodan** [comment]
I have this setup in use as you see it here right now. However if others could test this as well and report success/failure back here I'd sleep a little more easily.

File: contrib/systemd-with-workers/README.md [hunk 1]
```diff
@@ -0,0 +1,36 @@
+# Setup Synapse with Workers and Systemd
+
+This is a setup for using managing synapse with systemd including support for managing workers. It provides a `matrix-synapse-homeserver`, as well as a `matrix-synapse-worker@` service for any workers you require. Additionally to group the required services it sets up a `matrix.target`. You can of course add any bot- or bridge-services to this target as well by setting the respective `WantedBy` value in the `[Install]` section of their unit files to `matrix.target`.
```
Comment (richvdh) [inline contrib/systemd-with-workers/README.md]: please can you wrap your paragraphs at 80 characters?

File: contrib/systemd-with-workers/system/matrix-synapse-homeserver.service [hunk 2]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Synapse Matrix Homeserver
+
+[Service]
+Type=forking
```
Comment (richvdh) [inline contrib/systemd-with-workers/system/matrix-synapse-homeserver.service]: I'd recommend using `Type=simple` (and making sure that synapse is not configured to `daemonize`). Systemd is better at managing daemonization than synapse, and it will solve your problem with the pid files.

Comment (targodan) [reply contrib/systemd-with-workers/system/matrix-synapse-homeserver.service]: Does this work with workers?

All worker tutorials I have found so far seem to suggest that daemonizing is required. Or do you mean only non-daemonize the homeserver app and still daemonize the workers?

File: contrib/systemd-with-workers/README.md (hunk 1 above)
Comment (targodan) [reply contrib/systemd-with-workers/README.md]: Of course. Will do this afternoon/evening.

File: contrib/systemd-with-workers/system/matrix-synapse-homeserver.service (hunk 2 above)
Comment (targodan) [reply contrib/systemd-with-workers/system/matrix-synapse-homeserver.service]: Nevermind. I just tested it and it works. Will change everything to a non-forking setup.

**Comment by targodan** [comment]
All requested changes implemented.

**Comment by targodan** [comment]
Maybe wait a bit before merging still. Using a target you can't easily restart all services. I found [these possible solutions though](https://unix.stackexchange.com/questions/212438/how-to-stop-all-units-belonging-to-the-same-target) which are using either `StopWhenUnneeded` or `PartOf` in order to accomplish this.

I'll experiment with this soon and update the pull request with whichever solution I found easier to use/cleaner.

**Comment by targodan** [comment]
There we go. Now you can restart synapse and all its workers in one go. Even more, if the homeserver app crashes for any reason all workers are automatically stopped by systemd.

See the [README](https://github.com/targodan/synapse/blob/systemd-workers/contrib/systemd-with-workers/README.md#usage) for more information on what you can do with this setup.

From my point of view this is ready to merge.

I don't know about the buildkite fails. If you want me to rebase my branch against your current master just let me know.

**Comment by targodan** [comment]
I rebased against master in hopes to solve the checks. Don't know what's up with that.

**Comment by richvdh** [comment]
sorry, this fell off my radar a bit. Can you rebase against current develop instead?

**Comment by targodan** [comment]
Yes of course, sorry.
Just done that. :)

**Comment by targodan** [comment]
Actually found the problem. It's my changelog.d file. Fixing now.

File: contrib/systemd-with-workers/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
+# Setup Synapse with Workers and Systemd
+
+This is a setup for using managing synapse with systemd including support for
```
Comment (richvdh) [inline contrib/systemd-with-workers/README.md]:
```suggestion
This is a setup for managing synapse with systemd including support for
```

File: contrib/systemd-with-workers/README.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+systemctl enable matrix-synapse-worker@federation_writer.service
+systemctl restart matrix-synapse.service
+```
+
+## The Configs
+
+Make sure the `worker_app` is set in the `homeserver.yaml` and it does not fork.
+
+```
+worker_app: synapse.app.homeserver daemonize: false
```
Comment (richvdh) [inline contrib/systemd-with-workers/README.md]:
```suggestion
worker_app: synapse.app.homeserver 
daemonize: false
```

File: contrib/systemd-with-workers/README.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+
+Whichever configuration you choose, after the setup the following will enable
+automatically starting (and potentially restarting) your bot/bridge with the
+`matrix.target`.
+
+```
+systemctl enable <yourBotOrBridgeName>.service
+```
+
+**Note** that from an inactive synapse the bots/bridges will only started with
```
Comment (richvdh) [inline contrib/systemd-with-workers/README.md]:
```suggestion
**Note** that from an inactive synapse the bots/bridges will only be started with
```

File: contrib/systemd-with-workers/system/matrix.target [hunk 6]
```diff
@@ -0,0 +1,7 @@
+[Unit]
+Description=Contains matrix services like synapse, bridges and bots
+After=network.target
+AllowIsolate=no
```
Comment (richvdh) [inline contrib/systemd-with-workers/system/matrix.target]: isn't this the default? Is there any reason to add it here?

File: contrib/systemd-with-workers/system/matrix-synapse-worker@.service [hunk 7]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Synapse Matrix Worker
+After=matrix-synapse.service
+BindsTo=matrix-synapse.service
+
+[Service]
+Type=forking
```
Comment (richvdh) [inline contrib/systemd-with-workers/system/matrix-synapse-worker@.service]: shouldn't this be `simple` ?

Comment (targodan) [reply contrib/systemd-with-workers/system/matrix-synapse-worker@.service]: Yes, must have slipped through when I was copying stuff. Fixing now...

Comment (targodan) [reply contrib/systemd-with-workers/system/matrix-synapse-worker@.service]: Fixed

File: contrib/systemd-with-workers/system/matrix.target (hunk 6 above)
Comment (targodan) [reply contrib/systemd-with-workers/system/matrix.target]: Assuming you are refering to the `AllowIsolate=no`. Yes, it's the default value according to the [systemd docs](https://www.freedesktop.org/software/systemd/man/systemd.unit.html#AllowIsolate=). However I personally am a friend of doing important things explicitly. The matrix.target is one you would never want to isolate as all non matrix services would be stopped, effectively killing the system.

Thus I deem this important. Take the (admittedly unlikely) case that systemd decides to change the default value. We would still want this to be false/no as this target is designed as a grouping of services rather than an isolatable target.

To be fair even with this set to true an admin would have to do `systemctl isolate matrix.target` in order to cause mayhem, but I'd like to prevent this under any circumstances.

TL; DR: I deem it important that isolated is false and I personally prefer to do important things explicitly rather than implicitly by default values. But yes, it is the default value and I'll leave the final decision up to you. I won't be offended if you tell me to remove it. :slightly_smiling_face:

Comment (targodan) [reply contrib/systemd-with-workers/system/matrix.target]: If you feel like this warrants a deeper/longer discussion feel free to pm me in the matrix [@luca:targodan.de](https://matrix.to/#/@luca:targodan.de)

Comment (richvdh) [reply contrib/systemd-with-workers/system/matrix.target]: > However I personally am a friend of doing important things explicitly

I think you could argue that there are a bunch of other default settings which are just as important, and pretty soon you end up with a unit file a mile long. 

However, I don't have strong feelings here.

File: contrib/systemd-with-workers/system/matrix-synapse-worker@.service [hunk 8]
```diff
@@ -0,0 +1,17 @@
+[Unit]
+Description=Synapse Matrix Worker
+After=matrix-synapse.service
+BindsTo=matrix-synapse.service
```
Comment (jo-so) [inline contrib/systemd-with-workers/system/matrix-synapse-worker@.service]: Does the *BindsTo* enforce the *matrix-synapse.service* gets started, too? This seems to be bad when a worker is running on a host different from the Synapse service. Or do I have to start the Synapse process on each host running any worker?
